File: 1293.shortest-path-in-a-grid-with-obstacles-elimination.py
#
# @lc app=leetcode id=1293 lang=python3
#
# [1293] Shortest Path in a Grid with Obstacles Elimination
#

# @lc code=start
class Solution:
    def shortestPath(self, grid: List[List[int]], k: int) -> int:
        M, N = len(grid), len(grid[0])
        target = (M-1, N-1)

        def manhattan_distance(row: int, col: int) -> int:
            return target[0]-row+target[1]-col
        
        # (row, col, remain_eliminations)
        state = (0, 0, k)

        # (estimation, steps, state)
        # h(n) = manhattan_distance, g(n) = steps
        queue = [(manhattan_distance(0, 0), 0, state)]
        seen = set([state])

        while queue:
            estimation, steps, (row, col, remain_eliminations) = heapq.heappop(queue)

            # we can reach the target in the shortest path (manhattan distance),
            #   even if the remaining steps are all obstacles
            if estimation - steps <= remain_eliminations:
                return estimation
            
            # explore the four directions in the next step
            for new_row, new_col in [(row, col+1), (row+1, col), (row, col-1), (row-1, col)]:
                if (0 <= new_row < M) and (0 <= new_col < N):
                    new_eliminations = remain_eliminations - grid[new_row][new_col]
                    new_state = (new_row, new_col, new_eliminations)

                    if new_eliminations >= 0 and new_state not in seen:
                        seen.add(new_state)
                        new_estimation = manhattan_distance(new_row, new_col)+steps+1
                        heapq.heappush(queue, (new_estimation, steps+1, new_state))

        return -1



# @lc code=end

File: 1293.shortest-path-in-a-grid-with-obstacles-elimination.cpp
/*
 * @lc app=leetcode id=1293 lang=cpp
 *
 * [1293] Shortest Path in a Grid with Obstacles Elimination
 */

// @lc code=start
class Solution {
public:
    vector<int> dirs{-1, 0, 1, 0, -1};
    using State = vector<int>;
    int shortestPath(vector<vector<int>>& grid, int k) {
        int M = grid.size(), N = grid[0].size();
        vector<vector<vector<int>>> steps(M, vector<vector<int>>(N, vector<int>(k+1, INT_MAX)));

        queue<State> q;
        q.push({0, 0, k});
        steps[0][0][k] = 0;

        while(!q.empty()) {
            auto curr = q.front();
            q.pop();

            int row = curr[0], col = curr[1], rk = curr[2];

            if(row == M-1 and col == N-1) {
                return steps[row][col][rk];
            }

            for(int t=0; t<4; ++t) {
                int nr = row+dirs[t], nc = col+dirs[t+1];
                if(nr < 0 or nr >= M or nc < 0 or nc >= N) continue;
                int nk = rk - grid[nr][nc];
                if(nk >= 0 and steps[nr][nc][nk] == INT_MAX) {
                    steps[nr][nc][nk] = steps[row][col][rk] + 1;
                    q.push({nr, nc, nk});
                }
            }
        }

        return -1;

    }

};
// @lc code=end

